### PR TITLE
Path: don't ask for `Path.home` on expand if not needed

### LINKED ISSUE #8126: Path.expand evaluates Path.home even when no ~ is in the path
### Problem

`Path.expand` evaluates `Path.home` even when it's not needed, which results in this slightly odd behavior:
```
$ crystal eval <<< 'require "uri"; puts File.expand_path("/")'
/
$ unset HOME
$ crystal eval <<< 'require "uri"; puts File.expand_path("/")'
Missing ENV key: "HOME" (KeyError)
```

I discovered this because systemd's `DynamicUser` makes `HOME` be unset.

### Reason

`Path.expand` takes `Path.home` as a default argument: https://github.com/crystal-lang/crystal/blob/36a7557630d696acfe3496adcfdd17286d0e0393/src/path.cr#L560

which will evaluate `Path.home`, which in turn looks up `HOME`:
https://github.com/crystal-lang/crystal/blob/36a7557630d696acfe3496adcfdd17286d0e0393/src/path.cr#L949-L950

Version: master (36a7557630d696acfe3496adcfdd17286d0e0393)

### PR DESCRIPTION
**Pull request by asterite**
Fixes #8126

### CODE REVIEW

File: src/path.cr [hunk 1]
```diff
@@ -568,9 +569,9 @@ struct Path
     name = @name
```
Comment (Sija) [inline src/path.cr]:
```suggestion
     name = @name
     home ||= Path.home
```

Comment (ysbaddaden) [reply src/path.cr]: Wouldn't that force `Path.home` to always be evaluated? which is precisely what we want to avoid here?

Comment (Sija) [reply src/path.cr]: Ouch, you're right 👍

**Comment by straight-shoota** [comment]
I guess it's fine to use `nil` as a value for lazy `Path.home` evaluation. But it's maybe not very intuitive. In #7903 we discussed making `nil` mean to disable home path expansion.

**Pull request merged**